src/stash.rs

```diff
@@ -0,0 +1,287 @@
+use {raw, panic, StashFlags, Oid, Error, Repository, Signature, StashApplyProgress};
+use std::ffi::{CString, CStr};
+use util::{Binding};
+use libc::{c_int, c_char, size_t, c_void, c_uint};
+use build::{CheckoutBuilder};
+use std::mem;
+
+/// Stash application progress notification function.
+///
+/// Return `true` to continue processing, or `false` to
+/// abort the stash application.
+pub type StashApplyProgressCb<'a> = FnMut(StashApplyProgress) -> bool + 'a;
+
+/// This is a callback function you can provide to iterate over all the
+/// stashed states that will be invoked per entry.
+pub type StashCb<'a> = FnMut(usize, &str, &Oid) -> bool + 'a;
+
+#[allow(unused)]
+/// Stash application options structure
+pub struct StashApplyOptions<'cb> {
+    progress: Option<Box<StashApplyProgressCb<'cb>>>,
+    raw_opts: raw::git_stash_apply_options
+}
+
+impl<'cb> StashApplyOptions<'cb> {
+    /// Creates a default set of merge options.
+    pub fn new() -> StashApplyOptions<'cb> {
+        let mut opts = StashApplyOptions {
+            progress: None,
+            raw_opts: unsafe { mem::zeroed() },
+        };
+        assert_eq!(unsafe {
+            raw::git_stash_apply_init_options(&mut opts.raw_opts, 1)
+        }, 0);
+        opts
+    }
+
+    /// Set stash application flag
+    pub fn flags(&mut self, flags: raw::git_stash_apply_flags) -> &mut StashApplyOptions<'cb> {
+        self.raw_opts.flags = flags;
+        self
+    }
+
+    /// Options to use when writing files to the working directory
+    pub fn checkout_options(&mut self, mut opts: CheckoutBuilder) -> &mut StashApplyOptions<'cb> {
```

Comment:
I think this may need to be stored internally to persist the callbacks stored inside of `opts`